### DISCUSSION TITLE
PPStructure 识别表格设置参数 merge_no_span_structure = False 报错

### DISCUSSION THREAD

**Discussion by jiangpeng853** [PaddlePaddle/PaddleOCR#14284] [Q&A]
![image](https://github.com/user-attachments/assets/d48fe24e-4bc4-4633-9e68-487593642dc6)

报错信息：
Traceback (most recent call last):
  File "C:\pyprojects\gui_gyfy\vocdata\table_ocr3.py", line 77, in <module>
    result = table_engine(img)
  File "C:\pyprojects\gui_gyfy\.venv\lib\site-packages\paddleocr\paddleocr.py", line 926, in __call__
    res, _ = super().__call__(img, return_ocr_result_in_table, img_idx=img_idx)
  File "C:\pyprojects\gui_gyfy\.venv\lib\site-packages\paddleocr\ppstructure\predict_system.py", line 163, in __call__
    res, table_time_dict = self.table_system(
  File "C:\pyprojects\gui_gyfy\.venv\lib\site-packages\paddleocr\ppstructure\table\predict_table.py", line 109, in __call__
    pred_html = self.match(structure_res, dt_boxes, rec_res)
  File "C:\pyprojects\gui_gyfy\.venv\lib\site-packages\paddleocr\ppstructure\table\matcher.py", line 66, in __call__
    dt_boxes, rec_res = self._filter_ocr_result(pred_bboxes, dt_boxes, rec_res)
  File "C:\pyprojects\gui_gyfy\.venv\lib\site-packages\paddleocr\ppstructure\table\matcher.py", line 199, in _filter_ocr_result
    y1 = pred_bboxes[:, 1::2].min()
IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed

**Comment by GreatV** [comment]
从错误信息和引用的相关问题来看，`PPStructure` 在 `merge_no_span_structure=False` 时发生的报错与输入数据的维度不匹配有关。以下是可能的原因分析和解决方案：

---

### 问题分析

1. **错误核心**
   - 报错信息 `IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed` 表示在操作数组时尝试访问了额外的维度。
   - 对应代码位置：
     ```python
     y1 = pred_bboxes[:, 1::2].min()
     ```
     这里 `pred_bboxes` 被认为是二维数组，但在实际运行时是 1 维的。

2. **潜在问题**
   - 这个错误通常与输入数据的预处理或模型预测结果有关。`pred_bboxes` 是从表格检测模型返回的边界框预测结果，而问题可能出现在：
     - 检测模型返回了空结果（`pred_bboxes` 为空或形状不正确）。
     - 输入图片的格式或内容导致模型未能正确识别表格。
     - 参数 `merge_no_span_structure=False` 导致了未处理的分支逻辑问题。

3. **可能的触发原因**
   - **模型不匹配**：所使用的 `table_model_dir` 模型可能不支持 `merge_no_span_structure=False` 的配置。
   - **输入问题**：输入图片的尺寸、格式或内容可能导致模型无法正常输出表格检测结果。
   - **版本兼容性问题**：PaddleOCR 的版本与 PaddlePaddle 或所使用的模型之间可能存在不兼容问题。

---

### 解决方案

#### 1. **检查模型与配置**
   - 确保使用的表格识别模型支持 `merge_no_span_structure` 参数。
   - 检查 `table_model_dir` 的路径是否指向正确的表格模型（例如 `ch_ppstructure_openatom_SLANetv2_infer`）。
   - 如果模型不支持该参数，可以尝试将 `merge_no_span_structure=True`，以避免触发未处理的逻辑。

#### 2. **输入图片预处理**
   - 确保输入图片是清晰的、高分辨率的表格图片。
   - 检查图片的尺寸是否符合模型的输入限制（默认 `det_limit_side_len=960`）。
   - 在调用 `table_engine(img)` 前，输出图片的形状以确认其格式正确：
     ```python
     print(img.shape)
     ```

#### 3. **禁用 `merge_no_span_structure`**
   - 如果功能需求允许，可以省略该参数或设置为默认值 `True`：
     ```python
     table_engine = PPStructure(merge_no_span_structure=True)
     ```

#### 4. **升级 PaddleOCR 和 PaddlePaddle**
   - 确保 PaddleOCR 和 PaddlePaddle 的版本是最新的，并相互兼容：
     ```bash
     pip install --upgrade paddlepaddle paddleocr
     ```

#### 5. **使用安全的参数组合**
   - 从报错中可以看到，`ch_ppstructure_openatom_SLANetv2_infer` 模型可能存在兼容性问题。尝试使用官方推荐的参数和模型：
     ```python
     table_engine = PPStructure(
         table_model_dir="path/to/ch_ppstructure_mobile_v2.0_SLANet_infer",
         merge_no_span_structure=True
     )
     ```

#### 6. **调试模型输出**
   - 在 `matcher.py` 中或修改代码以打印 `pred_bboxes` 的形状和内容，确认问题来源：
     ```python
     print(pred_bboxes)
     print(pred_bboxes.shape)
     ```

---

### 参考相关问题

1. [Issue #13887](https://github.com/PaddlePaddle/PaddleOCR/issues/13887)
   - **描述**：表格识别报错 `IndexError: too many indices for array: array is 1-dimensional, but 2 were indexed`。
   - **解决方案**：用户通过更换更合适的模型（`ch_ppstructure_mobile_v2.0_SLANet_infer`）解决了问题。

2. [Issue #14199](https://github.com/PaddlePaddle/PaddleOCR/issues/14199)
   - **描述**：表格识别模型 `ch_ppstructure_openatom_SLANetv2_infer` 输出为空。
   - **解决方案**：问题可能与模型或输入图片不匹配有关。

3. [Issue #14134](https://github.com/PaddlePaddle/PaddleOCR/issues/14134)
   - **描述**：PDF 表格检测不准确。
   - **解决方案**：尝试调整模型、参数以及输入图片的格式。

---

### 示例代码

以下是一个安全的示例代码，可以用来测试模型：

```python
from paddleocr import PPStructure
import cv2

# 模型路径
table_model_dir = "path/to/ch_ppstructure_mobile_v2.0_SLANet_infer"
layout_model_dir = "path/to/picodet_lcnet_x1_0_fgd_layout_table_infer"
det_model_dir = "path/to/ch_PP-OCRv4_det_infer"
rec_model_dir = "path/to/ch_PP-OCRv4_rec_infer"

# 初始化 PPStructure
table_engine = PPStructure(
    table_model_dir=table_model_dir,
    layout_model_dir=layout_model_dir,
    det_model_dir=det_model_dir,
    rec_model_dir=rec_model_dir,
    merge_no_span_structure=True  # 或省略此参数
)

# 加载图片
image_path = "path/to/your_table_image.jpg"
img = cv2.imread(image_path)

# 确认图片加载成功
if img is None:
    raise ValueError(f"图片加载失败: {image_path}")
print(f"图片形状: {img.shape}")

# 表格识别
result = table_engine(img)

# 输出结果
print(result)
```

---

### 总结

- **优先检查模型和参数是否匹配**。
- **使用推荐的模型（如 `ch_ppstructure_mobile_v2.0_SLANet_infer`）**。
- **确保输入图片的格式正确且清晰**。
- 如果问题仍未解决，可以在 PaddleOCR 的 GitHub 提交 Issue，附上更多细节。

---
*Response generated by 🤖 [feifei-bot](https://github.com/GreatV/feifei) | chatgpt-4o-latest*
